File: Connect 4 AI.py
from colorama import init, Fore, Style, Back
init()

def updateColour(colour):
    if colour == 1: return 0
    return 1

def printPosition(position):
    print(Back.WHITE, end = '')
    for i in range(6):
        for j in range(7):
            if position[i][j] == '': print(Fore.BLACK + "[ ]", end = '')
            else:
                print(Fore.BLACK + "[", end = '')
                if position[i][j] == 1:
                    print(Fore.RED +'O', end = '')
                else:
                    print(Fore.BLUE + 'O', end = '')
                print(Fore.BLACK + "]", end = '')
        print()
    print(Fore.BLACK + " 0  1  2  3  4  5  6 ")
    print(Style.RESET_ALL, end = '')

def getUserMove(position, colour):
    #Full input error checking
    printPosition(position)
    while True:
        move = input("Enter your move: ")
        try:
            move = int(move)
            if move > 6:
                print("Please enter a valid column number between 0-6")
                continue
            if position[0][move] != '':
                print("Column is full, Please enter a valid column number")
                continue
            break
        except:
            print("Please enter a valid column number between 0-6")
    return updatePosition(position, move, colour)

def updatePosition(position, move, colour):
    #Takes in a position and adds a checker in the 'move' column
    emptySpaces = [position[1][move], position[2][move], position[3][move], position[4][move], position[5][move]].count('')
    position[emptySpaces][move] = colour
    return position

def decideNextMove(position, colour, depth = 5):
    #1. call genNextMoves on position
    #2. for each move generated call genPositions
    #3. analyse the loss rate of each move using analysePosition
    #4. choose the move with the least loss rate
    
    moves = genNextMoves(position, colour)
    moveLossRate = []*len(moves)
    lossColour = updateColour(colour)
    
    for i in range(len(moves)):
        #analyse the lossRate of a move
        positions = genPositions(moves[i], colour, depth)
        losses = 0
        other = 0
        for k in range(len(positions)):
            result = analysePosition(positions[k])
            if result == lossColour: losses = losses + 1
            else: other = other + 1
        moveLossRate.append(losses/(losses + other))
    
    move = moveLossRate.index(min(moveLossRate))
    #must column correct the move to take into account starting positions with full rows who did not analyse 7 moves
    i = 0
    while i <= move:
        if position[0][i] != '': move += 1
        i += 1
    return move

def genPositions(position, colour, depth = 5):
    #Repeatedly call genNextMoves until we reach the depth
    positions = [[position]]
    
    #Loop through and add positions as the next element in array (array looks like this [positions of depth0, positions of depth1, positions of depth2, ...])
    for i in range(depth):
        positions.append(genNextMoves(positions[i][0], colour))
        for k in range(1, len(positions[i])):
            positions[i+1].extend(genNextMoves(positions[i][k], colour))
        colour = updateColour(colour)
    return positions[depth]

def genNextMoves(position, colour):
    #returns an array of the positions possible in the next move
    #completedColumns avoids creating duplicate positions i.e. adding a checker in c1 then c2 is the same as c2 then c1
    positions = []
    for i in range(7):
        if position[0][i] != '': continue #check if top row is full
        
        positions.append([position[0][:], position[1][:], position[2][:], position[3][:], position[4][:], position[5][:]]) #Append a copy of position that is not linked to position
        positions[-1] = updatePosition(positions[-1], i, colour) #add a checker in column i
    return positions

def analysePosition(position):
    #take in a multidimensional array of a position and determine whether it is won for either side 0 = won for Red 1 = won for Blue 2 = even
    #1. check rows
    fullRows = 6
    for j in range(0,6):
        if position[j].count('') == 7: fullRows -= 1
        elif position[j].count(0) > 3:
            if connected(position[j], 0): return 0
        elif position[j].count(1) > 3:
            if connected(position[j], 1): return 1
            
    #2. check columns
    for k in range(0,7):
        column = []
        for l in range(5, 5 - fullRows, -1): column.append(position[l][k])
        if column.count(0) > 3:
            if connected(column, 0): return 0
        elif column.count(1) > 3:
            if connected(column, 1): return 1
            
    #3. check diagonals, I'm repeating a codeblock here instead of using a function because it is more efficient
    if fullRows < 4: return 2 #can't have 4 in a row if less than 4 rows
    
    #check top left (going down) diagonals
    for m in range(0,3):
        diagonal = []
        for n in range(6-fullRows, 6):
            if (n-m) < 0: continue
            diagonal.append(position[n][n - m])
        if diagonal.count(0) > 3:
            if connected(diagonal, 0): return 0
        elif diagonal.count(1) > 3:
            if connected(diagonal, 1): return 1
            
    for q in range(1,4):
        if (7-q) - (6 - fullRows) < 4: break #if diagonals get shorter than 4 checkers we don't have to check them
        diagonal = []
        for r in range(6 - fullRows, 7 - q): diagonal.append(position[r][r+q])
        if diagonal.count(0) > 3:
            if connected(diagonal, 0): return 0
        elif diagonal.count(1) > 3:
            if connected(diagonal, 1): return 1
            
    #check top right (going down) diagonals
    for o in range(0,3):
        diagonal = []
        for p in range(6-fullRows, 6):
            if (p-o) < 0: continue
            diagonal.append(position[p][6 - (p-o)])
        if diagonal.count(0) > 3:
            if connected(diagonal, 0): return 0
        elif diagonal.count(1) > 3:
            if connected(diagonal, 1): return 1
            
    for s in range(1,4):
        if (7-s) - (6 - fullRows) < 4: break
        diagonal = []
        for t in range(6 - fullRows, 7 - s): diagonal.append(position[t][6 - (s + t)])
        if diagonal.count(0) > 3:
            if connected(diagonal, 0): return 0
        elif diagonal.count(1) > 3:
            if connected(diagonal, 1): return 1                         
    return 2

def connected(line, colour):
    #checks if 4 of the same colour checkers are connected
    #fully optimised, might look bad but this function is run millions of times so it needs to be as fast as possible
    i = 0
    length = len(line) - 3
    while i < length:
        if line[i] == colour:
            if line[i+1] == colour:
                if line[i+2] == colour:
                    if line[i+3] == colour: return True
                    return False
                i += 1
            i += 1
        i += 1
    return False


samplePosition = [[1 ,'','','','','',''],
                  [0 ,'','','','','',''],
                  [1 ,1 ,'','','',1 ,1 ],
                  [0 ,0 ,1 ,'','',1 ,1 ],
                  [0 ,1 ,1 ,1 ,0 ,0 ,0 ],
                  [0 ,0 ,0 ,1 ,0 ,1 ,1 ]]

samplePosition2 = [[1,'','',0 ,'','',''],
                   [1,'','',1 ,'','',''],
                   [0,'','',1 ,'','',''],
                   [1,'','',0 ,'',1 ,''],
                   [1,'','',0 ,'',1 ,''],
                   [1,'','',0 ,'',1 ,'']]

startPosition = [['','','','','','',''],
                ['','','','','','',''],
                ['','','','','','',''],
                ['','','','','','',''],
                ['','','','','','',''],
                ['','','','','','','']]

def mainLine(position):
    print("Welcome to connect 4! Input your moves by entering the column number you wish to place a checker in, columns are numbered 0 to 6 left to right")
    colour = 1
    while position[0].count('') != 0: #while game is not over
        position = getUserMove(position, colour)
        printPosition(position)
        #check if game won
        if analysePosition(position) == colour:
            print("--------You win!--------")
            return
        
        colour = updateColour(colour)
        move = decideNextMove(position, colour)
        position = updatePosition(position, move, colour)
        #check if game won
        if analysePosition(position) == colour:
            print("--------You lose :(--------")
            return
        
        colour = updateColour(colour)

#mainLine(startPosition)
n = input("Press enter to quit")

File: connect4ai.java
package connect4ai;

import java.util.ArrayList;
import java.util.Scanner;

class Analysis{
    //contains all the analysis methods used by the program
    
    //checks if a line (maximum 7 checkers) contains 4 checkers in a row, returns 0 if red wins, 1 if blue wins, 2 if none
    public static int connected(int[] line){        
        int j = 0;
        int maxTests = line.length - 3;
        while (j < maxTests){
            if (line[j+1] == line[j]){
                if (line[j+2] == line[j]){
                    if (line[j+3] == line[j]){
                        //note if there are 4 connected empty spots it is impossible for there to be another
                        //4 checkers in a row, so returning 2 here is fine
                        return line[j];
                    }
                    j++;
                }
                j++;
            }
            j++;
        }
        return 2;
    }
    
    //Analyse a position created from an even parent (only have to check the modified row, column and diagonals)
    public static int analysePosition(int[][] board, int[] move){
        //check row
        int result = connected(board[move[0]]);
        if (result != 2){ return result;}
        
        //check column
        int[] column = new int[6];
        for(int i = 0; i < 6; i++){
            column[i] = board[i][move[1]];
        }
        result = connected(column);
        if (result != 2){ return result;}
        
        //check top left to bottom right diagonal
        int[] topLeftStart = new int[2]; //coordinates of the start of the diagonal
        int length;
        if (move[0] >= move[1]){
            topLeftStart[0] = move[0] - move[1];
            topLeftStart[1] = 0;
            length = 6 - topLeftStart[0];
        }
        else { 
            topLeftStart[0] = 0;
            topLeftStart[1] = move[1] - move[0];
            length = 7 - topLeftStart[1];
        }
        
        //construct diagonal
        int[] diagonal = new int[length];
        int i = 0;
        while (i < length){
            diagonal[i] = board[topLeftStart[0] + i][topLeftStart[1] + i];
            i++;
        }
        
        result = connected(diagonal);
        if (result != 2){ return result;}
        
        //check top right to bottom left diagonal
        int[] topRightStart = new int[2]; //coordinates of the start of the diagonal
        if (move[0] >= 6 - move[1]){
            topRightStart[0] = move[0] - (6 - move[1]);
            topRightStart[1] = 6;
            length = 6 - topRightStart[0];
        }
        else{
            topRightStart[0] = 0;
            topRightStart[1] = move[1] + move[0];
            length = topRightStart[1] + 1;
        }
        
        //construct diagonal array
        i = 0;
        int[] diagonal2 = new int[length];
        while (i < length){
            diagonal2[i] = board[topRightStart[0] + i][topRightStart[1] - i];
            i++;
        }
        return connected(diagonal2);
    }
    
    //Heuristic to analyse even positions by number of adjacent pieces 
    public static int leafNodeValue(Position pos, int opponentColour){
        int callerScore = 0;
        int opponentScore = 0;
        int sum;
        
        for(int i = 5; i > 0; i--){ //loop through rows from bottom to top
            //stop searching if the row is empty
            sum = 0;
            for (int j = 0; j < 7; j++){
                sum += pos.board[i][j];
            }
            if (sum == 14){return callerScore - opponentScore/2;}
            
            for(int k = 0; k < 6; k++){ //search all columns except the last, so that their is always a column to the right
                //compare to pieces directly above
                if (pos.board[i][k] == pos.board[i-1][k]){
                    if (pos.board[i][k] == opponentColour){opponentScore++;}
                    else if (pos.board[i][k] != 2){callerScore++;}
                }
                
                //compare to pieces diagonally above and to the right
                if (pos.board[i][k] == pos.board[i-1][k+1]){
                    if (pos.board[i][k] == opponentColour){opponentScore++;}
                    else if (pos.board[i][k] != 2){callerScore++;}
                }
                
                //compare to pieces diagonally above and to the left
                if (pos.board[i][k+1] == pos.board[i-1][k]){
                    if (pos.board[i][k+1] == opponentColour){opponentScore++;}
                    else if (pos.board[i][k+1] != 2){callerScore++;}
                }
                
                //compare to pieces to the right
                if (pos.board[i][k] == pos.board[i][k+1]){
                    if (pos.board[i][k] == opponentColour){opponentScore++;}
                    else if (pos.board[i][k] != 2){callerScore++;}
                }
            }
            
            //check the last column
            if (pos.board[i][6] == pos.board[i-1][6]){
                    if (pos.board[i][6] == opponentColour){opponentScore++;}
                    else if (pos.board[i][6] != 2){callerScore++;}
            }
        }
        
        return callerScore - opponentScore/2; //favours gaining adjacent pieces over stopping your opponent
    }
}

class Position{
    //whether the position is won or lost: -100 = loss, 0 = even, 100 = won
    public int state;
    //represents the board as a multidimensional array, 0 = red checker 1 = blue checker 2 = empty
    public int[][] board = new int[6][7];
    //the players colour
    public static int playerColour;
    //references to children positions, useful for the minimax algorithm
    public ArrayList<Position> children = new ArrayList<>();
    public int minChildColumn = 0; //the smallest column a checker could be placed in to create a new child
    private int numChildrenStore = -1; //storage for the number of children, only calculated through numChildren() method
    //useful for minimax algorithm, between -100 and 100
    public int nodeValue;
    //level of the position, needed for iterative deepening
    public int level;
    
    
    //explicitly define a position (only used for startingPosition or sample positions)
    public Position(int[][] array, int result){
        for(int i = 0; i < 6; i++){
            System.arraycopy(array[i], 0, board[i], 0, array[i].length);
        }
        if (result == 2){state = 0;}
        else if (result == playerColour){state = -100;}
        else{state = 100;}
        level = 0;
    }
    
    //create a position by placing a checker in the parent position
    public Position(Position Parent, int[] move, int colour){
        for(int i = 0; i < 6; i++){
            System.arraycopy(Parent.board[i], 0, board[i], 0, Parent.board[i].length);
        }
        board[move[0]][move[1]] = colour;
        level  = Parent.level;
        level++;
        
        int result = Analysis.analysePosition(board, move);
        if (result == 2){state = 0;}
        else if (result == playerColour){state = -100;}
        else{state = 100;}
    }
    
    public void displayPosition(){
        for (int[] board1 : board) {
            for (int j = 0; j < board1.length; j++) {
                switch (board1[j]) {
                    case 0: System.out.printf("[O]");
                    break;
                    case 1: System.out.printf("[X]");
                    break;
                    case 2: System.out.printf("[ ]");
                    break;
                }
            }
            System.out.println();
        }
        System.out.println(" 0  1  2  3  4  5  6");
    }
    
    public int numChildren(){
        if (numChildrenStore != -1){return numChildrenStore;}
        
        int numChildren = 7;
        //check if there are any full columns
        for(int i = 0; i < 7; i++){
            if (board[0][i] != 2){numChildren--;}
        }
        
        numChildrenStore = numChildren; //stores numChildren for future reference
        return numChildren;
    }
}

class gameTree{
    //the root of the gameTree
    public Position root;
    //the largest size a level can be without taking too long to analyse
    private final int levelLimit = 3000000;
    //keeps track of how many levels findBestMove has analysed
    private int currentLevelSize;
    private int depth;
    
    private int updateColour(int colour){
        if (colour == 0){return 1;}
        return 0;
    }
    
    //adds a child to the parent position by placing a checker in minColumn or the next free column after that
    public void addChild(Position parent, int colour){
        //find column to place checker in
        while (parent.board[0][parent.minChildColumn] != 2){ //while column is full
            parent.minChildColumn++;
        }
        
        //create new position
        int lowestEmptySpace = 5;
        while (parent.board[lowestEmptySpace][parent.minChildColumn] != 2){
            lowestEmptySpace--;
        }
        int[] move = {lowestEmptySpace, parent.minChildColumn};
        Position newPos = new Position(parent, move, colour);
        
        parent.children.add(newPos);
        parent.minChildColumn++;
        currentLevelSize++;
    }
    
    public gameTree(Position startPosition){
        root = startPosition;
    }
    
    private int maximiserValue(Position pos, int alpha, int beta){
        //stop recursion if it is won or lost
        if (pos.state != 0){return pos.state;}
        //stop recursion if it is a leaf node
        if (pos.level == depth){return Analysis.leafNodeValue(pos, updateColour(Position.playerColour));}
        
        pos.nodeValue = -100; //initialise nodeValue to worst case scenario
        int numChildren = pos.numChildren();
        for(int i = 0; i < numChildren; i++){
            //generate child if needed
            if (pos.children.size() == i){
                addChild(pos, updateColour(Position.playerColour));
            }
            
            int childValue = minimiserValue(pos.children.get(i), alpha, beta); //find the nodeValue of a child
            
            //if childValue is better than beta the minimiser would never go here
            if (childValue > beta){return childValue;}
            //update nodeValue and alpha if a better option is available
            if (childValue > pos.nodeValue){
                pos.nodeValue = childValue;
                if (childValue > alpha){alpha = childValue;}
            }
        }
        return pos.nodeValue;
    }
    
    private int minimiserValue(Position pos, int alpha, int beta){
        //stop recursion if it is won or lost
        if (pos.state != 0){return pos.state;}
        //stop recursion if it is a leaf node
        if (pos.level == depth){return Analysis.leafNodeValue(pos, Position.playerColour);}
        
        pos.nodeValue = 100; //initialise nodeValue to worst case scenario
        int numChildren = pos.numChildren();
        for(int i = 0; i < numChildren; i++){
            //generate child if needed
            if (pos.children.size() == i){
                addChild(pos, Position.playerColour);
            }
            
            int childValue = maximiserValue(pos.children.get(i), alpha, beta); //find the nodeValue of a child
            
            //if childValue is worse than alpha the maximiser would never go here
            if (childValue < alpha){return childValue;}
            //update nodeValue and beta if a better option is available
            if (childValue < pos.nodeValue){
                pos.nodeValue = childValue;
                if (childValue < beta) {beta = childValue;}
            }
        }
        return pos.nodeValue;
    }
    
    public int findBestMove(){
        root.nodeValue = -100; //initialise nodeValue to worst case scenario
        int bestMove = 0;
        int numChildren = root.numChildren();
        
        //note alpha = best already explored option along the path to the root for the maximiser and beta is the same for the minimiser
        for(int i = 0; i < numChildren; i++){
            //generate child if needed
            if (root.children.size() == i){
                addChild(root, updateColour(Position.playerColour));
            }
            
            int childValue = minimiserValue(root.children.get(i), root.nodeValue, 100);
            
            //update nodeValue if a better option is found
            if (childValue > root.nodeValue) {
                root.nodeValue = childValue;
                bestMove = i;
            }
            if (root.nodeValue == 100) {break;}
        }
        return bestMove;
    }
    
    //iterative deepening algorithm that uses findBestMove
    public Position decideMove(){
        int bestMove = 0;
        int newMove;
        depth = 1;
        
        //keep searching until the next level would have a larger size than the levelLimit
        while (currentLevelSize*5 < levelLimit){
            currentLevelSize = 0;
            
            //update bestMove and depth
            newMove = findBestMove();
            System.out.println("Best Move: " + newMove + " Evaluation: " + root.nodeValue + " Depth: " + depth + " Level size: " + currentLevelSize);
            
            if (root.nodeValue == -100){break;} //if the algorithm thinks the position is lost it won't bother finding a solution, so return the previous bestMove
            bestMove = newMove;
            depth++;
            
            //if there are no more positions to generate we have reached max depth and are done
            if (currentLevelSize == 0){break;}
        }
        return root.children.get(bestMove);
    }
}

class Connect4ai {
    private static Scanner sc = new Scanner(System.in);
    
    public static void main(String[] args) {
        Position.playerColour = 0;
        int[][] startPosArray = {{2,2,2,2,2,2,2}, 
                                 {2,2,2,2,2,2,2},
                                 {2,2,2,2,2,2,2},
                                 {2,2,2,2,2,2,2},
                                 {2,2,2,2,2,2,2},
                                 {2,2,2,1,2,2,2}};
              
        int[][] samplePosArray = {{1,2,2,2,2,2,2},
                                  {0,1,2,2,2,2,0},
                                  {1,1,2,2,2,1,1},
                                  {0,0,1,2,1,0,1},
                                  {0,1,1,0,0,1,0},
                                  {0,1,0,0,0,1,1}};
        
        int[][] testPosArray = {{2,2,2,2,2,2,2}, 
                                {2,2,2,2,2,2,2},
                                {2,2,2,2,2,2,2},
                                {2,2,2,0,0,0,2},
                                {2,2,2,2,0,1,1},
                                {2,2,2,2,0,1,1}};
                
        Position startPos = new Position(startPosArray, 2);
        Position samplePos = new Position(samplePosArray, 2);
        Position testPos = new Position(testPosArray, 2);
        
        String Continue;
        do{
            gameLoop(startPos);
            System.out.printf("Continue playing? [Y/N] ");
            Continue = sc.next();
            System.out.println();
        } while (!"N".equals(Continue));
        
    }
    
    private static void gameLoop(Position startPos){
        gameTree tree = new gameTree(startPos);
        Position currentPos = startPos;
        
        System.out.println("Welcome to connect4! Enter your moves by typing a column number between 0 and 6");
        System.out.println("The system evaluates positions from -100 to 100 with 100 being best for the computer");
        currentPos.displayPosition();
        while (true){
            currentPos = getPlayerMove(currentPos);
            currentPos.displayPosition();
            if (currentPos.state == -100){
                System.out.println("You win!");
                break;
            }
            
            tree.root = currentPos;
            currentPos = tree.decideMove();
            currentPos.displayPosition();
            if (currentPos.state == 100){
                System.out.println("You lose");
                break;
            }
        }
    }
    
    private static Position getPlayerMove(Position pos){
        System.out.printf("Enter move: ");
        int column = sc.nextInt();
        System.out.println();
        
        int lowestEmptySpace = 5;
        while (pos.board[lowestEmptySpace][column] != 2){
            lowestEmptySpace--;
        }
        
        int[] move = {lowestEmptySpace, column};
        Position playerPos = new Position(pos, move, Position.playerColour);
        playerPos.level = 0;
        return playerPos;
    }
}
